File: daily_problems/2026/08/0805/personal_submission/cf106632l_yawn_sean.py
# Submission link: https://codeforces.com/gym/106632/submission/385366226
def main():
    class LinkedList:
        def __init__(self, n):
            self.head = 0
            self.tail = 1
            
            self.nxt = [-1] * (n + 5)
            self.prv = [-1] * (n + 5)
            self.tags = [0] * (n + 5)
            
            self.nxt[0] = 1
            self.prv[1] = 0
            
            self.pos = {}
            self.sz = 0
            self.flg = False
            self.pt = 2
        
        def check(self, idx):
            return idx in self.pos
        
        def size(self):
            return self.sz
        
        def insert(self, idx):
            if self.flg:
                self.prv[self.pt] = self.head
                self.nxt[self.pt] = self.nxt[self.head]
            else:
                self.prv[self.pt] = self.prv[self.tail]
                self.nxt[self.pt] = self.tail
    
            self.nxt[self.prv[self.pt]] = self.pt
            self.prv[self.nxt[self.pt]] = self.pt
            
            self.tags[self.pt] = idx
            self.pos[idx] = self.pt
            
            self.pt += 1
            self.sz += 1
        
        def delete(self, idx):
            p = self.pos[idx]
            del self.pos[idx]
            
            self.prv[self.nxt[p]] = self.prv[p]
            self.nxt[self.prv[p]] = self.nxt[p]
            
            self.sz -= 1
        
        def get_head(self):
            return self.tags[self.prv[self.tail] if self.flg else self.nxt[self.head]]
        
        def reverse(self):
            self.flg = not self.flg
        
        def output(self):
            ans = []
            
            if self.flg:
                cur = self.tail
                while self.prv[cur] != self.head:
                    cur = self.prv[cur]
                    ans.append(self.tags[cur])
            else:
                cur = self.head
                while self.nxt[cur] != self.tail:
                    cur = self.nxt[cur]
                    ans.append(self.tags[cur])
            
            return ans
    
    t = II()
    outs = []
    
    rnd = random.getrandbits(30)
    
    for _ in range(t):
        n, m, k = MII()
        
        idxs = [0] * n
        states = [0] * n
        
        for i in range(n):
            idxs[i], states[i] = MII()
            idxs[i] ^= rnd
        
        reverses = [0] * n
        if k:
            for idx in GMI():
                reverses[idx] = 1
        
        total = LinkedList(n)
        broken = LinkedList(n)
        
        for i in range(n):
            if total.check(idxs[i]):
                total.delete(idxs[i])
                if broken.check(idxs[i]):
                    broken.delete(idxs[i])
                
                total.insert(idxs[i])
                if not states[i]:
                    broken.insert(idxs[i])
            
            elif total.size() < m:
                total.insert(idxs[i])
                if not states[i]:
                    broken.insert(idxs[i])
            
            elif broken.size():
                idx = broken.get_head()
                total.delete(idx)
                broken.delete(idx)
                
                total.insert(idxs[i])
                if not states[i]:
                    broken.insert(idxs[i])
            
            elif states[i]:
                idx = total.get_head()
                total.delete(idx)
                
                total.insert(idxs[i])
                if not states[i]:
                    broken.insert(idxs[i])
            
            if reverses[i]:
                total.reverse()
                broken.reverse()
    
        ans = total.output()
        outs.append(str(len(ans)))
        outs.append(' '.join(str(x ^ rnd) for x in ans))
    
    print('\n'.join(outs))

File: daily_problems/2026/08/0805/personal_submission/cf106632l_yawn_sean.cpp
#include <bits/stdc++.h>
// #include "atcoder/all"
// #pragma GCC optimize("Ofast")
// #pragma GCC optimize("unroll-loops")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx2,tune=native")

using namespace std;

auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

struct ListNode {
	public:
		explicit ListNode(int n) {
			head = 0;
			tail = 1;

			nxt = vector<int>(n + 5, -1);
			prv = vector<int>(n + 5, -1);
			tag = vector<int>(n + 5, 0);

			nxt[0] = 1;
			prv[1] = 0;

			sz = 0, flg = false, pt = 2;
		}

		bool check(int idx) {return pos.find(idx) != pos.end();}

		int size() {return sz;}

		void insert(int idx) {
			if (flg) {
				prv[pt] = head;
				nxt[pt] = nxt[head];
			}
			else {
				prv[pt] = prv[tail];
				nxt[pt] = tail;
			}

			nxt[prv[pt]] = pt;
			prv[nxt[pt]] = pt;

			tag[pt] = idx;
			pos[idx] = pt;

			pt ++, sz ++;
		}

		void remove(int idx) {
			int p = pos[idx];
			pos.erase(idx);

			prv[nxt[p]] = prv[p];
			nxt[prv[p]] = nxt[p];

			sz --;
		}

		int get_head() {return tag[flg ? prv[tail] : nxt[head]];}

		void reverse() {flg = !flg;}

		void output(int rnd) {
			cout << sz << '\n';

			if (flg) {
				int cur = tail;
				while (prv[cur] != head) {
					cur = prv[cur];
					cout << (tag[cur] ^ rnd) << ' ';
				}
			}
			else {
				int cur = head;
				while (nxt[cur] != tail) {
					cur = nxt[cur];
					cout << (tag[cur] ^ rnd) << ' ';
				}
			}

			cout << '\n';
		}

	private:
		int head, tail, sz, pt;
		bool flg;
		vector<int> nxt, prv, tag;
		unordered_map<int, int> pos;
};

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int t, rnd = rng();
	cin >> t;

	while (t --) {
		int n, m, k;
		cin >> n >> m >> k;

		vector<int> idxs(n), states(n);

		for (int i = 0; i < n; i ++) {
			cin >> idxs[i] >> states[i];
			idxs[i] ^= rnd;
		}

		vector<int> reverses(n, 0);
		while (k --) {
			int idx; cin >> idx;
			reverses[idx - 1] = 1;
		}

		auto total = ListNode(n);
		auto broken = ListNode(n);

		for (int i = 0; i < n; i ++) {
			if (total.check(idxs[i])) {
				total.remove(idxs[i]);
				if (broken.check(idxs[i])) broken.remove(idxs[i]);
				total.insert(idxs[i]);
				if (!states[i]) broken.insert(idxs[i]);
			}
			else if (total.size() < m) {
				total.insert(idxs[i]);
				if (!states[i]) broken.insert(idxs[i]);
			}
			else if (broken.size()) {
				int idx = broken.get_head();
				total.remove(idx);
				broken.remove(idx);
				total.insert(idxs[i]);
				if (!states[i]) broken.insert(idxs[i]);
			}
			else if (states[i]) {
				int idx = total.get_head();
				total.remove(idx);
				total.insert(idxs[i]);
				if (!states[i]) broken.insert(idxs[i]);
			}

			if (reverses[i]) total.reverse(), broken.reverse();
		}

		total.output(rnd);
	}

	return 0;
}